File: src/roomba500.py
#!/usr/bin/env python
import time

from sensorframe import SensorFrame

def twoComplementOf( value ):
    """ returns two bytes (ints) in high, low order
    whose bits form the input value when interpreted in
    two's complement
    """
    # if positive or zero, it's OK
    if value >= 0:
        eqBitVal = value
    # if it's negative, I think it is this
    else:
        eqBitVal = (1<<16) + value

    return ( (eqBitVal >> 8) & 0xFF, eqBitVal & 0xFF )

def _bitOfByte( bit, byte ):
    """ returns a 0 or 1: the value of the 'bit' of 'byte' """
    if bit < 0 or bit > 7:
        print('Your bit of', bit, 'is out of range (0-7)')
        print('returning 0')
        return 0
    return ((byte >> bit) & 0x01)

def twosComplementInt2bytes( highByte, lowByte ):
    """ returns an int which has the same value
    as the twosComplement value stored in
    the two bytes passed in

    the output range should be -32768 to 32767

    chars or ints can be input, both will be
    truncated to 8 bits
    """
    # take everything except the top bit
    topbit = _bitOfByte( 7, highByte )
    lowerbits = highByte & 127
    unsignedInt = lowerbits << 8 | (lowByte & 0xFF)
    if topbit == 1:
        # with sufficient thought, I've convinced
        # myself of this... we'll see, I suppose.
        return unsignedInt - (1 << 15)
    else:
        return unsignedInt

class Roomba500:

    def __init__(self, ser, fullRotationInSensorTicks, ticksPerCm, robotWheelDistanceInCm):
        self.ser = ser
        self.fullRotationInSensorTicks = fullRotationInSensorTicks
        self.ticksPerCm = ticksPerCm
        self.robotWheelDistanceInCm = robotWheelDistanceInCm
        self.lastBumperRight = False
        self.lastBumperLeft = False
        self.lastRightWheelDropped = False
        self.lastLeftWheelDropped = False
        self.lastKnownReferencePose = None
        self.leftWheelDistance = 0
        self.rightWheelDistance = 0
        self.mainbrushPWM = 0
        self.sidebrushPWM = 0
        self.vacuumPWM = 0

    def safeMode(self):
        self.ser.write(chr(128))
        self.ser.write(chr(131))

    def passiveMode(self):
        self.ser.write(chr(128))

    def updateMotorControl(self):
        self.ser.write(chr(144))
        self.ser.write(chr(self.mainbrushPWM & 0xff))
        self.ser.write(chr(self.sidebrushPWM & 0xff))
        self.ser.write(chr(self.vacuumPWM & 0xff))

    def playNote(self, note, duration):
        self.ser.write(chr(140))
        self.ser.write(chr(0))
        self.ser.write(chr(1))
        self.ser.write(chr(note))
        self.ser.write(chr(duration))
        self.ser.write(chr(141))
        self.ser.write(chr(0))

    def drive(self, speedLeft, speedRight):
        leftSpeedHigh, leftSpeedLow = twoComplementOf(speedLeft)
        rightSpeedHigh, rightSpeedLow = twoComplementOf(speedRight)
        self.ser.write(chr(145))
        self.ser.write(chr(rightSpeedHigh))
        self.ser.write(chr(rightSpeedLow))
        self.ser.write(chr(leftSpeedHigh))
        self.ser.write(chr(leftSpeedLow))

    def readSensorFrame(self):
        self.ser.write(chr(149))
        self.ser.write(chr(12))
        self.ser.write(chr(43))
        self.ser.write(chr(44))
        self.ser.write(chr(25))
        self.ser.write(chr(26))
        self.ser.write(chr(7))

        self.ser.write(chr(46))
        self.ser.write(chr(47))
        self.ser.write(chr(48))
        self.ser.write(chr(49))
        self.ser.write(chr(50))
        self.ser.write(chr(51))

        self.ser.write(chr(35))

        leftWheel = ord(self.ser.read()) << 8 | ord(self.ser.read())
        rightWheel = ord(self.ser.read()) << 8 | ord(self.ser.read())
        batteryCharge = ord(self.ser.read())  << 8 | ord(self.ser.read())
        batteryCapacity = ord(self.ser.read()) << 8 | ord(self.ser.read())
        bumperState = ord(self.ser.read())

        lightBumpLeft = ord(self.ser.read()) << 8 | ord(self.ser.read())
        lightBumpFrontLeft = ord(self.ser.read()) << 8 | ord(self.ser.read())
        lightBumpCenterLeft = ord(self.ser.read()) << 8 | ord(self.ser.read())
        lightBumpCenterRight = ord(self.ser.read()) << 8 | ord(self.ser.read())
        lightBumpFrontRight = ord(self.ser.read()) << 8 | ord(self.ser.read())
        lightBumpRight = ord(self.ser.read()) << 8 | ord(self.ser.read())

        oimode = ord(self.ser.read())

        result = SensorFrame()
        result.leftWheel = leftWheel
        result.rightWheel = rightWheel
        result.batteryCharge = batteryCharge
        result.batteryCapacity = batteryCapacity
        result.bumperState = bumperState
        result.lightBumperLeft = lightBumpLeft
        result.lightBumperFrontLeft = lightBumpFrontLeft
        result.lightBumperCenterLeft = lightBumpCenterLeft
        result.lightBumperCenterRight = lightBumpCenterRight
        result.lightBumperFrontRight = lightBumpFrontRight
        result.lightBumperRight = lightBumpRight
        result.oimode = oimode

        return result

File: src/roomba500.cpp
#include <stdio.h>
#include <string>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>
#include <stdexcept>
#include <sys/ioctl.h>
#include "sensorframe.cpp"

class Roomba500 {

    private:
        int fd;

    public:
        bool lastBumperRight;
        bool lastBumperLeft;
        bool lastRightWheelDropped;
        bool lastLeftWheelDropped;
        char mainbrushPWM;
        char sidebrushPWM;
        char vacuumPWM;

    Roomba500(std::string device, int baudrate) {
        // Open and initialize serial interface
        // Code taken from https://www.cmrr.umn.edu/~strupp/serial.html
        //this->fd = open(device.c_str(), O_RDWR | O_NOCTTY | O_NDELAY);
        this->fd = open(device.c_str(), O_RDWR | O_NOCTTY | O_NDELAY);
        if (this->fd == -1) {
            throw std::invalid_argument("cannot open serial interface");
        } else {
            // Read blocking behavior
            fcntl(this->fd, F_SETFL, 0);
        }

        // Set serial port baud rate
        struct termios options;
        tcgetattr(fd, &options);

        if (baudrate == 115200) {
            cfsetispeed(&options, B115200);
            cfsetospeed(&options, B115200);
        } else {
            throw std::invalid_argument("unsupported baud rate");
        }

        options.c_cflag     |= (CLOCAL | CREAD);
        options.c_lflag     &= ~(ICANON | ECHO | ECHOE | ISIG);
        options.c_oflag     &= ~OPOST;
        options.c_cc[VMIN]  = 1; // Blocking read, as min 1 bytes in required in input buffer
        options.c_cc[VTIME] = 10;

        options.c_cflag &= ~PARENB;
        options.c_cflag &= ~CSTOPB;
        options.c_cflag &= ~CSIZE;
        options.c_cflag |= CS8;
        options.c_iflag &= ~(IXON | IXOFF | IXANY);

        // Configure interface
        tcsetattr(fd, TCSANOW, &options);

        this->lastBumperRight = false;
        this->lastBumperLeft = false;
        this->lastRightWheelDropped = false;
        this->lastLeftWheelDropped = false;
        this->mainbrushPWM = 0;
        this->sidebrushPWM = 0;
        this->vacuumPWM = 0;
    }

    void safeMode() {

        unsigned char data[] = {128, 131};

        int written = write(fd, &data, sizeof(data));
        if (written != sizeof(data)) {
            throw std::invalid_argument("error sending safeMode command");
        }
        tcdrain(fd);
    }

    void passiveMode() {
        unsigned char data[] = {128};
        int written = write(fd, &data, sizeof(data));
        if (written != sizeof(data)) {
            throw std::invalid_argument("error sending passiveMode command");
        }
        tcdrain(fd);
    }

    void updateMotorControl() {
        unsigned char data[] = {131, 144, (unsigned char) (mainbrushPWM & 0xff), (unsigned char) (sidebrushPWM & 0xff), (unsigned char) (vacuumPWM & 0xff)};
        int written = write(fd, &data, sizeof(data));
        if (written != sizeof(data)) {
            throw std::invalid_argument("error updating motor control");
        }
        tcdrain(fd);
    }

    void playNote(unsigned char note, unsigned char duration) {
        unsigned char data[] = {131, 140, 0, 1, note, duration, 141, 0};
        int written = write(fd, &data, sizeof(data));
        if (written != sizeof(data)) {
            throw std::invalid_argument("error playing note");
        }
        tcdrain(fd);
    }

    void drive(int speedLeft, int speedRight) {

        unsigned int speedLeftUnsigned;
        if (speedLeft < 0) {
            speedLeftUnsigned = (1<<16) + speedLeft;
        } else {
            speedLeftUnsigned = speedLeft;
        }

        unsigned int speedRightUnsigned;
        if (speedRight < 0) {
            speedRightUnsigned = (1<<16) + speedRight;
        } else {
            speedRightUnsigned = speedRight;
        }

        unsigned char data[] = {131, 145, (unsigned char)((speedRightUnsigned >> 8) & 0xff), (unsigned char)(speedRightUnsigned & 0xff),
                                     (unsigned char)((speedLeftUnsigned >> 8) & 0xff), (unsigned char)(speedLeftUnsigned & 0xff)};
        int written = write(fd, data, sizeof(data));
        if (written != sizeof(data)) {
            throw std::invalid_argument("error sending drive command");
        }
        tcdrain(fd);
    }

    unsigned char readData() {
        unsigned char result;
        int readBytes = read(fd, &result, 1);
        return result;
    }

    void set_port_read_min_size(int min_size) {
        struct termios settings;

        if (tcgetattr(fd, &settings)) {
           /* handle error */
            throw std::invalid_argument("cannot read settings for serial interface");
        }

        /* set the minimimum no. of chracters to read in each
         * read call.
         */
        settings.c_cc[VMIN]  = min_size;
        /* set “read timeout between characters” as 100 ms.
         * read returns either when the specified number of chars
         * are received or the timout occurs */
        settings.c_cc[VTIME] = 1; /* 1 * .1s */

        if (tcsetattr (fd, TCSANOW, &settings)) {
            throw std::invalid_argument("cannot write settings for serial interface");
        }
    }

    SensorFrame readSensorFrame() {

        char rawdata[23] __attribute((aligned(32)));

        tcflush(fd, TCIFLUSH);

        unsigned char data[] = {131, 149, 13, 43, 44, 25, 26, 7, 46, 47, 48, 49, 50, 51, 35, 45};
        int written = write(fd, &data, sizeof(data));
        if (written != sizeof(data)) {
            throw std::invalid_argument("sending read sensors command");
        }
        tcdrain(fd);

        // Wait for response or timeout
        set_port_read_min_size(23);

        int bytesAvailable = 0;
        ioctl(fd, FIONREAD, &bytesAvailable);

        SensorFrame sensorFrame = SensorFrame();

        int recvbytes = 0;
        int maxfd = fd + 1;
        int index = 0;
        /* set the timeout as 1 sec for each read */
        struct timeval timeout = {0, 30 * 1000}; // 30ms timeout
        fd_set readSet;

        FD_ZERO(&readSet);
        FD_SET(fd, &readSet);
        if (select(maxfd, &readSet, NULL, NULL, &timeout) > 0) {
            if (FD_ISSET(fd, &readSet)) {
                recvbytes = read(fd, &rawdata, sizeof(rawdata));
                if(recvbytes) {
                    sensorFrame.leftWheel = rawdata[0] << 8 | rawdata[1]; // packet 43
                    sensorFrame.rightWheel = rawdata[2] << 8 | rawdata[3]; // packet 44
                    sensorFrame.batteryCharge = rawdata[4] << 8 | rawdata[5]; // packet 25
                    sensorFrame.batteryCapacity = rawdata[6] << 8 | rawdata[7]; // packet 26
                    sensorFrame.bumperState = rawdata[8]; // packet 7
                    sensorFrame.lightBumperLeft = rawdata[9] << 8 | rawdata[10]; // packet 46
                    sensorFrame.lightBumperFrontLeft = rawdata[11] << 8 | rawdata[12]; // packet 47
                    sensorFrame.lightBumperCenterLeft = rawdata[13] << 8 | rawdata[14]; // packet 48
                    sensorFrame.lightBumperCenterRight = rawdata[15] << 8 | rawdata[16]; // packet 49
                    sensorFrame.lightBumperFrontRight = rawdata[17] << 8 | rawdata[18]; // packet 50
                    sensorFrame.lightBumperRight = rawdata[19] << 8 | rawdata[20]; // packet 51
                    sensorFrame.oimode = rawdata[21]; // packet 35
                    sensorFrame.lightBumperStat = rawdata[22]; // packet 45
                } else {
                    throw std::invalid_argument("failed to read data");
                }

            }
        } else {
            /* select() – returns 0 on timeout and -1 on error condtion */
            throw std::invalid_argument("timeout or error while waiting for data");
        }

        return sensorFrame;
    }

    ~Roomba500() {
        close(fd);
    }
};